### PR TITLE
[TPU] Add support for Pytorch XLA FSDP

### LINKED ISSUE #13209: Feature request: FSDP native strategy for TPUs
A recent contribution to the pytorch_xla repo allows using FSDP in PyTorch XLA for sharding Module parameters across data-parallel workers. https://github.com/pytorch/xla/pull/3431.

### PR DESCRIPTION
**Pull request by gkroiz**
## What does this PR do?
Address feature request #13209 to implement FSDP in Pytorch XLA for TPUs. Use the `XLAFSDPStrategy` to use FSDP on TPUs.

Fixes #13209 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [X] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [X] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by gkroiz** [comment]
cc @Liyang90 for visibility

**Comment by gkroiz** [comment]
When looking at the [xla_fsdp documentation](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py), checkpoints are saved as sharded checkpoints, which can be combined into a single checkpoint using [consolidate_sharded_model_checkpoints](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/state_dict_utils.py#L109). To save the sharded checkpoints, the user will need to specify device-specific paths (i.e, for 4 devices, 4 separate paths). [Here](https://github.com/pytorch/xla/blob/master/test/test_train_mp_mnist_fsdp_with_ckpt.py) is a good reference example of how to save checkpoints and consolidate.

This is the current idea I have:
```
    def save_checkpoint(
        self, path: _PATH, state: Dict[str, Union[Module, Optimizer, Any]], storage_options: Optional[Any] = None
    ) -> None:
        """Save model, optimizer, and other state as a checkpoint file.

        Args:
            path: A path to where the file(s) should be saved
            state: A dictionary with contents to be saved. If the dict contains modules or optimizers, their
                state-dict will be retrieved and converted automatically.
            storage_options: Additional options for the ``CheckpointIO`` plugin
        """
        rank_zero_warn("Saving checkpoints in the XLAFSDPStrategy requires saving a sharded checkpoint for each device. \
        Please make sure the path specified is device specific!")

        state = self._convert_stateful_objects_in_state(state)

        if 'model' not in state:
             raise ValueError('XLAFSDPStrategy requires the saved state to include \'model\'.')
        if 'shard_metadata' not in state:
             raise ValueError('XLAFSDPStrategy requires the saved state to include \'shard_metadata\'.')
        if 'optimizer' not in state
             raise ValueError('XLAFSDPStrategy requires the saved state to include \'optimizer\'.')

        self.checkpoint_io.save_checkpoint(state, path, storage_options=storage_options)
```

Would we also want to combine the sharded checkpoints in `save_checkpoint()`? My only concern here is that the use would need to specify several other function arguments like [`ckpt_prefix`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/state_dict_utils.py#L117-L118), [`ckpt_suffix`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/state_dict_utils.py#L119-L125), and [`save_path`](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/state_dict_utils.py#L126-L129). Combining sharded checkpoints is beneficial as the end result is a single checkpoint that can be loaded in for inference etc. The downside is that this strategy's `save_checkpoint()` will require several more work from the user to use the correct function arguments.

What are everyone's thoughts? If we do not include checkpoint consolidation directly in the `save_checkpoint()` function, we can also (1) create a separate function that does this, or (2) reference some documentation on the matter. 

cc @carmocca  @Liyang90

**Comment by carmocca** [comment]
@gkroiz We should try to stay as close to PyTorch's FSDP implementation. For example, see Fabric's FSDP implementation which is our most up-to-date version at the moment: https://github.com/Lightning-AI/lightning/blob/a5c43d3b2b80f5fc769d7ed0ea511c0bd6733c6b/src/lightning/fabric/strategies/fsdp.py#L306-L368
Here, the user specifies the directory where the shards are saved and the shard names are managed automatically by the strategy or an internal.

XLA's FSDP implementation was forked from the original fairscale implementation. The PyTorch version is quickly evolving, so it would be useful to know what's the plan for XLA's. For instance, XLA's doesn't offer the `use_orig_params` argument or the new saving APIs. This means unifying might not be easy (so it's not a hard requirement), but if we do it then it should be easier to update if XLA's version is updated to match in the future.

For consolidation, this would be done separately and as an opt-in. We are in the process of deciding the desired design for this in Lightning (cc @awaelchli): Our goal is to support {saving,loading} {sharded,consolidated} checkpoints. But to limit the scope of this work we can focus on saving and loading just sharded versions. See for example this comment: https://github.com/Lightning-AI/lightning/issues/16526#issuecomment-1503541511

File: src/lightning/fabric/strategies/xla_fsdp.py [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+from lightning.fabric.strategies.strategy import _BackwardSyncControl, TBroadcast
+from lightning.fabric.utilities.imports import _TORCH_GREATER_EQUAL_2_0
+from lightning.fabric.utilities.rank_zero import rank_zero_only, rank_zero_warn
+from lightning.fabric.utilities.types import _PATH, ReduceOp
+
+if TYPE_CHECKING and _XLA_AVAILABLE:
+    from torch_xla.distributed.parallel_loader import MpDeviceLoader
+
+
+class XLAFSDPStrategy(ParallelStrategy):
```
Comment (carmocca) [inline src/lightning/fabric/strategies/xla_fsdp.py:42]: nit: I suggest `FSDPXLAStrategy` instead as "FSDP" is more informative than "XLA" hence it should go first

**Comment by gkroiz** [comment]
Thank you for https://github.com/Lightning-AI/lightning/pull/17421#issuecomment-1520382873, I agree that we want to stay as close as possible to PyTorch's FSDP implementation but difficulties arise since the XLA FSDP implementation is somewhat behind. I'll make some adjustments to try to match https://github.com/Lightning-AI/lightning/blob/a5c43d3b2b80f5fc769d7ed0ea511c0bd6733c6b/src/lightning/fabric/strategies/fsdp.py#L306-L368 and then we can further adjust from there

File: examples/fabric/tpu/train_fsdp_mnist.py
[the remark this answers is not in the document]
Comment (gkroiz) [reply examples/fabric/tpu/train_fsdp_mnist.py:1]: Sounds good, I can remove this. I used this moreso for my own sanity check comparing model weights when using xla fsdp since there were not any code examples of Fabric + xla AFAIK.

File: src/lightning/fabric/strategies/xla_fsdp.py (hunk 1 above)
Comment (gkroiz) [reply src/lightning/fabric/strategies/xla_fsdp.py:42]: I originally had `FSDPXLAStrategy` when I first opened the PR but the [xla source code](https://github.com/pytorch/xla/blob/master/torch_xla/distributed/fsdp/xla_fully_sharded_data_parallel.py) uses `XLAFSDPStrategy` so I switched to `XLAFSDPStrategy` to stay consistent. Does this context change anything? If not we can change back to `FSDPXLAStrategy`.

[the remark this answers is not in the document]
Comment (gkroiz) [reply src/lightning/fabric/strategies/xla_fsdp.py:42]: Thanks for the suggestion, this would remove some overlap. I will make the change

File: src/lightning/fabric/strategies/xla_fsdp.py
[the remark this answers is not in the document]
Comment (gkroiz) [reply src/lightning/fabric/strategies/xla_fsdp.py:1]: Added in this PR

File: src/lightning/fabric/connector.py [hunk 2]
```diff
@@ -523,11 +529,11 @@ def _lazy_init_strategy(self) -> None:
 
         # TODO: should be moved to _check_strategy_and_fallback().
         # Current test check precision first, so keep this check here to meet error order
-        if isinstance(self.accelerator, XLAAccelerator) and not isinstance(
-            self.strategy, (SingleDeviceXLAStrategy, XLAStrategy)
+        if isinstance(self.accelerator, TPUAccelerator) and not isinstance(
```
Comment (carmocca) [inline src/lightning/fabric/connector.py:532]: There was a mistake merging master here

File: src/lightning/fabric/connector.py [hunk 3]
```diff
@@ -523,11 +529,11 @@ def _lazy_init_strategy(self) -> None:
...
+            self.strategy, (SingleTPUStrategy, XLAStrategy, XLAFSDPStrategy)
```
Comment (carmocca) [inline src/lightning/fabric/connector.py:533]: Since you are subclassing XLAStrategy, adding XLAFSDP here is not necessary

File: src/lightning/pytorch/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -159,6 +160,11 @@ def __init__(
         # 4. Instantiate Strategy - Part 1
         if self._strategy_flag == "auto":
             self._strategy_flag = self._choose_strategy()
+
+        # Change fsdp to xla_fsdp if using TPU
+        if self._strategy_flag == "fsdp" and self._accelerator_flag == "tpu":
+            self._strategy_flag = "xla_fsdp"
```
Comment (carmocca) [inline src/lightning/pytorch/trainer/connectors/accelerator_connector.py:166]: This logic would fit well inside the `_check_strategy_and_fallback` method

File: src/lightning/fabric/connector.py [hunk 5]
```diff
@@ -154,6 +155,11 @@ def __init__(
         # 4. Instantiate Strategy - Part 1
         if self._strategy_flag == "auto":
             self._strategy_flag = self._choose_strategy()
+
+        # Change fsdp to xla_fsdp if using TPU
+        if self._strategy_flag == "fsdp" and self._accelerator_flag == "tpu":
+            self._strategy_flag = "xla_fsdp"
```
Comment (carmocca) [inline src/lightning/fabric/connector.py:161]: Same suggestion

**Review by carmocca — COMMENTED** [review]
This is looking great. I'll do a thorough pass this week

File: src/lightning/pytorch/trainer/connectors/accelerator_connector.py (hunk 4 above)
Comment (gkroiz) [reply src/lightning/pytorch/trainer/connectors/accelerator_connector.py:166]: Good idea, I will make the change

**Comment by gkroiz** [comment]
Want to check in here to make sure this PR does not become too stale and that it eventually lands. Other than updating these changes with master, what else is needed?

**Comment by gkroiz** [comment]
Due to the size of this PR, I think it is best to split this into two separate PRs, one for Fabric and one for Trainer.

**Comment by gkroiz** [comment]
Fabric support for Pytorch XLA FSDP on TPUs: #18126
